kybra/compiler/kybra_generate/src/canister_method/post_upgrade_method/build.rs

```diff
@@ -13,64 +16,121 @@ use crate::{
 
-        let (params, guard_function_name) =
-            if let Some(post_upgrade_function_def) = post_upgrade_function_def_option {
-                let (guard_function_name, params, return_type) = (
-                    post_upgrade_function_def
-                        .get_guard_function_name()
-                        .map_err(Error::into),
-                    post_upgrade_function_def.build_params(InternalOrExternal::Internal),
-                    post_upgrade_function_def.build_return_type(),
-                )
-                    .collect_results()?;
+    Ok((post_upgrade_params, guard_function_name))
+}
 
-                if !canister_method::is_void(return_type) {
-                    return Err(ReturnTypeMustBeVoid::err_from_stmt(
-                        post_upgrade_function_def,
-                        CanisterMethodType::PostUpgrade,
-                    )
-                    .into());
-                }
+fn select_post_upgrade_method_params(
+    init_params: Vec<Param>,
+    post_upgrade_params: Vec<Param>,
+) -> Vec<Param> {
+    if post_upgrade_params.is_empty() {
+        init_params
+    } else {
+        post_upgrade_params
+    }
+}
 
-                (params, guard_function_name)
-            } else {
-                (vec![], None)
-            };
+fn generate_post_upgrade_method_body(
+    entry_module_name: &String,
+    init_function_def: Option<&SourceMapped<&Located<StmtKind>>>,
+    post_upgrade_function_def: Option<&SourceMapped<&Located<StmtKind>>>,
+) -> Result<TokenStream, Vec<Error>> {
+    rust::generate(
+        init_function_def,
+        post_upgrade_function_def,
+        entry_module_name,
+    )
+}
+
+fn handle_multiple_function_defs(
+    function_defs: &Vec<SourceMapped<&Located<StmtKind>>>,
+    canister_method_type: CanisterMethodType,
+) -> Result<(), Vec<Error>> {
+    if function_defs.len() > 1 {
+        Err(MultipleSystemMethods::err_from_stmt(&function_defs, canister_method_type).into())
+    } else {
+        Ok(())
+    }
+}
 
-        let call_to_post_upgrade_py_function = match &post_upgrade_function_def_option {
-            Some(post_upgrade_function_def) => {
-                post_upgrade_function_def.generate_call_to_py_function()?
-            }
-            None => quote!(),
-        };
+fn validate_post_upgrade_return_type(
+    post_upgrade_function_def: &SourceMapped<&Located<StmtKind>>,
+) -> Result<(), Vec<Error>> {
+    let return_type = post_upgrade_function_def.build_return_type()?;
 
-        Ok((
-            PostUpgradeMethod {
-                params: params.clone(),
-                body: rust::generate(&params),
-                guard_function_name,
-            },
-            call_to_post_upgrade_py_function,
-        ))
+    if !canister_method::is_void(return_type) {
+        Err(ReturnTypeMustBeVoid::err_from_stmt(
+            post_upgrade_function_def,
+            CanisterMethodType::PostUpgrade,
+        )
+        .into())
+    } else {
+        Ok(())
```

Comment:
Consider doing an early return rather than an if/else. I thought you preferred that.